server/src/main/java/io/crate/execution/dml/TranslogIndexer.java

```diff
@@ -0,0 +1,183 @@
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentFactory;
+import org.elasticsearch.index.mapper.IdFieldMapper;
+import org.elasticsearch.index.mapper.MapperParsingException;
+import org.elasticsearch.index.mapper.Mapping;
+import org.elasticsearch.index.mapper.ParsedDocument;
+import org.elasticsearch.index.mapper.SequenceIDFields;
+import org.elasticsearch.index.mapper.TextFieldMapper;
+import org.elasticsearch.index.mapper.Uid;
+
+import io.crate.expression.reference.doc.lucene.SourceParser;
+import io.crate.metadata.doc.DocSysColumns;
+import io.crate.metadata.doc.DocTableInfo;
+import io.crate.sql.tree.ColumnPolicy;
+import io.crate.types.DataType;
+
+/**
+ * Parses transaction log entries based on column information from a {@link DocTableInfo}
+ */
+public class TranslogIndexer {
+
+    private record ColumnIndexer<T>(DataType<T> dataType, ValueIndexer<Object> valueIndexer) {}
+
+    private final Map<String, ColumnIndexer<?>> indexers = new HashMap<>();
+    private final Map<String, List<String>> tableIndexSources = new HashMap<>();
+    private final boolean ignoreUnknownColumns;
+    private final SourceParser sourceParser;
+
+    /**
+     * Creates a new TranslogIndexer backed by a DocTableInfo instance
+     */
+    @SuppressWarnings("unchecked")
+    public TranslogIndexer(DocTableInfo table) {
+        for (var ref : table.columns()) {
+            var storageSupport = ref.valueType().storageSupport();
+            if (storageSupport != null) {
+                var columnIndexer = new ColumnIndexer<>(
+                    ref.valueType(),
+                    (ValueIndexer<Object>) storageSupport.valueIndexer(table.ident(), ref, table::getReference));
+                indexers.put(ref.column().name(), columnIndexer);
+            }
+        }
+        for (var ref : table.indexColumns()) {
+            for (var source : ref.columns()) {
+                tableIndexSources.computeIfAbsent(source.column().fqn(), _ -> new ArrayList<>()).add(ref.storageIdent());
+            }
+        }
+        this.ignoreUnknownColumns = table.columnPolicy() != ColumnPolicy.STRICT;
+        this.sourceParser = new SourceParser(table);
+    }
+
+    /**
+     * Convert a transaction log entry to a ParsedDocument to be indexed
+     * @param id        the document ID
+     * @param source    the transaction log entry bytes
+     */
+    public ParsedDocument index(String id, BytesReference source) {
+
+        Document doc = new Document();
+
+        var stream = new BytesStreamOutput();
+        try (XContentBuilder xContentBuilder = XContentFactory.json(stream)) {
+
+            Mapping dynamicUpdate = null;
+
+            try {
+                populateLuceneFields(source, doc, xContentBuilder);
+            } catch (MapperParsingException e) {
+                // TODO Can we throw a MappingUpdateRequired exception here instead?
+                dynamicUpdate = Mapping.EMPTY;
```

Comment:
I'm not sure either. Can we try it? Raising an error seems at least safer in that it would tell us if there's something wrong?

Reply:
It involves reworking a fair amount of the translog callers so I'd like to leave it for a followup if that's OK, just to keep this PR relatively simple.

Reply:
Sounds good